Points to improve 1/3 - Datepicker component

Ravaelles opened the pull request:
Based on ~true story~ [ this document](https://creativeforce.atlassian.net/wiki/spaces/DEV/pages/1351909447/Points+to+improve+Slawek+OKR+s+T30): a major refactor of Datepicker. Extracted number of child components and two mixins.

Datepicker now only stores the `date` and `dateInput` and controls subcomponents, [which have taken all its logic](https://github.com/tectonic/vue-bootstrap/blob/feature/points-to-improve-datepicker/src/Datepicker.vue#L23-L74).

think-01 commented on package.json:
```diff
@@ -29,6 +29,7 @@
     "karma-sourcemap-loader": "^0.3.8",
     "karma-webpack": "v4.0.2",
     "mocha": "^8.3.2",
+    "moment": "^2.18.1",
```
I'd rather add it to `peerDependencies` so that if embedding module has it it will nt get imported twice

think-01 commented on src/Datepicker/DatepickerControlButtons.vue:
```diff
@@ -0,0 +1,78 @@
+      this.$emit('changeDate', this.dateNow);
+      this.$emit('changeView', this.mode === 'time' ? 'clock' : 'calendar');
+    },
+    clearSelectedDate () {
```
is that even used?

think-01 commented on src/Datepicker/DatepickerControlButtons.vue:
```diff
@@ -0,0 +1,78 @@
+    clearSelectedDate () {
+      this.$emit('clearDate');
+    },
+    close () {
```
I'd prefer to have it in one convention i.e. emit strictly from template (prefered) of via method like here

think-01 commented on src/Datepicker/DatepickerHourMinute.vue:
```diff
@@ -0,0 +1,220 @@
+  },
+  methods: {
+    setHour (hour) {
+      const date = this.date || this.dateNow;
```
that could go into computed property so you can reuse it in other methods too

think-01 commented on src/Datepicker/DatepickerHourMinute.vue:
```diff
@@ -0,0 +1,220 @@
+
+      this.$emit('changeDate', newDate);
+    },
+    hours () {
```
why not just use:

```
data() {
    return {
      hours: [["00", "01", "02", "03"], ["04", "05", "06", "07"], ["08", "09", "10", "11"], ["12", "13", "14", "15"], ["16", "17", "18", "19"], ["20", "21", "22", "23"]],
      minutes: [["00", "05", "10", "15"], ["20", "25", "30", "35"], ["40", "45", "50", "55"]]

    };
  },
```
?

think-01 commented on src/Datepicker/DatepickerHourMinute.vue:
```diff
@@ -0,0 +1,220 @@
+    },
+    hoursFormatted (hours = null) {
+      if (hours === null) {
+        hours = this.date === null ? this.dateNow.getHours() : this.date.getHours();
```
`(this.date || this.dateNow).getHours()`
( could get extracted into computed as its used in multiple places )

think-01 commented on src/Datepicker/DatepickerHourMinute.vue:
```diff
@@ -0,0 +1,220 @@
+      return ('0' + value).slice(-2);
+    },
+    toggleAmPm () {
+      this.setClock('hours', this.date.getHours() < 12 ? 'increment' : 'decrement', 12);
```
what if `this.date === null`?

think-01 commented on src/Datepicker/DatepickerHourMinute.vue:
```diff
@@ -0,0 +1,220 @@
+      }
+      return this.pad(minutes);
+    },
+    pad (value) {
```
why making it a method? lets just do `const pad = value => ('0' + value).slice(-2);` outside of a component we don't need a component context in here

think-01 commented on src/Datepicker/DatepickerYearMonth.vue:
```diff
@@ -0,0 +1,201 @@
+      let year = this.years !== undefined ? this.years[0] : this.year - 5;
+      const key = direction === '+' ? 0 : 11;
+      while (true) {
+        let result = Array.from({ length: 12 }, (_, i) => eval(year + direction + i)).sort((a, b) => a - b);
```
omg thi is soooo ugly :/

`const result = [...Array(12).keys()].map( () => year + (direction === '+' ? 1 : -1))`

Ravaelles replied:
This part comes from latest task of Dimitrios or it may be even older:
https://github.com/tectonic/vue-bootstrap/pull/69/files#diff-1a4907c34ebb0ec2da25a59237a1da56023265e6fa22740058027bd88cf51f79R724

It may be ugly, but very concise and it works. It does a lot indeed.

think-01 replied:
@kontoulis @Ravaelles I will insist here - eval is evil, please get rid of it, for example in the way I showed above :)

kontoulis replied:
@think-01 Your version looks better indeed :)

think-01 commented on src/Datepicker/DatepickerYearMonth.vue:
```diff
@@ -0,0 +1,201 @@
+      }), 4);
+    },
+    translate (key) {
+      if (key in this.translations) {
```
can we use ternary here?

think-01 commented on src/Datepicker/date-formatter.js:
```diff
@@ -0,0 +1,92 @@
+      }
+    },
+    formatter (date) {
+      const format = (
```
why not have this as `computed` instead of `includesTime` and `includesDate`?

think-01 commented on src/Datepicker/date-formatter.js:
```diff
@@ -0,0 +1,92 @@
+        .format(format);
+    },
+
+    formatDateTime (date, forceDefaultFormat = false) {
```
I'm a lil bit confused about that one because it looks lika a bit of duplicate for `formatter` logic - not sure if I understand logic well but can we possibly somehow reduce it?

think-01 commented on src/Datepicker/date-formatter.js:
```diff
@@ -0,0 +1,92 @@
+      }
+
+      return date.getFullYear() +
+        '-' + this.pad(date.getMonth() + 1) +
```
since `pad` is used again maybe we should extract it to external module and import when necessary?

think-01 reviewed (changes requested):
nice annoying job @Ravaelles :+1:  a few comments left though

Ravaelles commented:
Ok @think-01, following your initial suggestion and Kazik's ready solution to extract moment.js into external dependency the task is ready for a final review. I addressed all issues you raised.

antennaio commented:
There are some new props (`timezones`, `momentLocale`, `momentDateFormat`, `momentTimeFormat`).

Please document the new props in `vuepress/components/datepicker.md`. @Ravaelles 

BTW, is there any way to give those props defaults to make them optional? If not, it's a breaking change, so this will be a major release (1.4.0?). If there are any breaking changes you know of please mention them in the change log:

https://github.com/tectonic/vue-bootstrap/blob/master/CHANGELOG.md

Ravaelles commented:
Thanks for the info @antennaio. I've documented the new props and I made them optional in component, so no breaking change was introduced. Falcon staging back to 🟢 

Waiting for final approval.
